File: algorithms/A/Advantage Shuffle/Advantage Shuffle.py
class Solution:
    def advantageCount(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # sort by ascend order
        nums1.sort()

        # heapq in python is a minimal heap, so if we need to use a maximal heap, the element in the queue should be like: (-num, index)
        q = []
        for i, n in enumerate(nums2):
            heapq.heappush(q, (-n, i))

        n = len(nums1)
        l = 0
        r = n - 1
        ans = [0] * n

        while q:
            # pop out the maximum number from nums2
            num, i = heapq.heappop(q)
            num = -num

            # no number in nums1 > max number in nums2, so grap the minimum numer from nums1
            if nums1[r] <= num:
                ans[i] = nums1[l]
                l += 1
            else:
                ans[i] = nums1[r]
                r -= 1

        return ans

File: scripts/algorithms/A/Advantage Shuffle/Advantage Shuffle.cpp
class Solution {
public:
    vector<int> advantageCount(vector<int>& nums1, vector<int>& nums2) {
        
        int n1 = nums1.size();
        
        int n2 = nums2.size();
        
        multiset<int> s(nums1.begin(), nums1.end());
        
        for(int i = 0; i < n2; i++)
        {
            int val = nums2[i];
            
            auto it = s.upper_bound(val);
            
            if(it != s.end())
            {
                nums1[i] = *it;
                
                s.erase(it);
            }
            else
            {
                nums1[i] = -1;
            }
        }
        
        for(int i = 0; i < n1; i++)
        {
            if(nums1[i] == -1)
            {
                nums1[i] = *s.begin();
                
                s.erase(s.begin());
            }
        }
        
        return nums1;
    }
};
